### ISSUE TITLE
Accessing BP callbacks in JS

### ISSUE THREAD

**Issue by Youdaman** [ncsoft/Unreal.js#47]
I have the following setup in BP that works:
![screenshot 2016-03-16 16 48 08](https://cloud.githubusercontent.com/assets/287330/13803349/fbb79fc4-eb97-11e5-96d4-5c1a88bbf646.png)

I'm attempting to perform the same operation in JS in my PlayerController subclass:
```
        getSteamFriends() {
            let PC = this
            let friendsProxy = GetFriendsCallbackProxy.GetAndStoreFriendsList(GWorld, PC)
            friendsProxy.Activate()
            let friends = PC.GetStoredFriendsList().FriendsList
//             let friends = [];
//             AdvancedFriendsLibrary.GetStoredFriendsList(PC, friends)
            return friends
        }
```

The C++ class definition is as follows:
```
UCLASS(MinimalAPI)
class UGetFriendsCallbackProxy : public UOnlineBlueprintCallProxyBase
{
	GENERATED_UCLASS_BODY()

	// Called when the friends list successfully was retrieved
	UPROPERTY(BlueprintAssignable)
	FBlueprintGetFriendsListDelegate OnSuccess;

	// Called when there was an error retrieving the friends list
	UPROPERTY(BlueprintAssignable)
	FBlueprintGetFriendsListDelegate OnFailure;

	// Gets the players list of friends from the OnlineSubsystem and returns it, can be retrieved later with GetStoredFriendsList
	UFUNCTION(BlueprintCallable, meta=(BlueprintInternalUseOnly = "true", WorldContext="WorldContextObject"), Category = "Online|AdvancedFriends")
	static UGetFriendsCallbackProxy* GetAndStoreFriendsList(UObject* WorldContextObject, class APlayerController* PlayerController);

	virtual void Activate() override;
```

If I delete the SteamFriends variable and the BP code then the JS does not work, so my understanding is that in JS I have to emulate the "OnSuccess" Exec from the BP in order for the subsquent GetStoredFriendsList to work.

I can see that the class has a BluePrintGetFriendsListDelegate OnSuccess but I'm not sure how to to use that. Is it like a callback in JS?

**Comment by Youdaman** [comment]
TL;DR

This works: `PC.GetStoredFriendsList().FriendsList`

But only after an initial fetch using GetAndStoreFriendsList and then setting a variable with the result.

If I break the link before the SET in the screenshot above then `PC.GetStoredFriendsList().FriendsList` stops working. I'm not sure why I have to set a variable for it to work. I tried just doing a Print String from the OnSuccess pin but that didn't work, so it seems in order for the list to be saved and available for the next fetch it needs to be stored in a BP variable.

I will ask the author of the plugin if there's any special magic I need to perform, however I think it's just my lack of knowledge regarding delegates and how they work so you might spot the problem :)

**Comment by arthursw** [comment]
When you implement the c++ version of the function, you can see that the c++ version is called when nothing is connected in the blueprint, and the blueprint version is called as soon as something is connected to it.

**Comment by Youdaman** [comment]
The author of the Steam plugin suggested I make use of the OnSuccess callback for the delegate, so I did this:
```
let friendsProxy = GetFriendsCallbackProxy.GetAndStoreFriendsList(GWorld, PC)
friendsProxy.OnSuccess = function(steamFriends) {
  //console.log(JSON.stringify(steamFriends,null,2));
  console.log('GREAT SUCCESS', steamFriends.length);
}
friendsProxy.Activate()
```
And this works (prints the number of Steam friends to console). However I think I've encountered a bug unrelated to Unreal.js in that I cannot remove the BP version of the call to GetAndStoreFriendsList otherwise things stop working (it doesn't seem to actually store friends unless it has an explicity call to a SET node in BP with its results -- even if you don't use that variable anywhere else, it just seems to need it in order for the internal saving of the friends list to trigger).

I believe I'm doing the delegate callback correctly, unless there's a better way to do this in Unreal.js?

My only worry is if my findings above are not caused by a bug in this plugin but due to me translating things into Unreal.js incorrectly or using the wrong syntax or semantics.

**Comment by Youdaman** [comment]
Closing this issue because I'm an idiot :P 

Turns out I had a dependency in my code on the variable I'd created in BP such that when I deleted the SET call in BP that caused the code to branch differently, not even calling the Steam functions.

I can confirm that the code block in the above comment works standalone without any BP stuff :)
